token/core/zkatdlog/crypto/rp/bulletproof.go

```diff
@@ -0,0 +1,422 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package rp
+
+import (
+	math "github.com/IBM/mathlib"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto/common"
+	"github.com/pkg/errors"
+)
+
+// RangeProof proves that a committed value < max
+type RangeProof struct {
+	// T1 is a Pedersen commitment to a random tau1
+	T1 *math.G1
+	// T2 is a Pedersen commitment a random tau2
+	T2 *math.G1
+	// Tau = tau1x + tau2x^2 for a random challenge x
+	Tau *math.Zr
+	// C is a hiding Pedersen commitment to vectors left = (b_0, ...,b_{n-1})
+	// and right = (b_0 - 1, ..., b_{n-1}-1) with randomness rho, where
+	// v = \sum_{i=0}^{n-1} b_i 2^i, and n = BitLength
+	C *math.G1
+	// D is a hiding Pedersen commitment to two random vectors
+	// with randomness eta
+	D *math.G1
+	// Delta = rho + x eta
+	Delta *math.Zr
+	// InnerProduct is the value of the inner product of the vectors committed in the non-hiding
+	// commitment C*D^x/P^Delta
+	InnerProduct *math.Zr
+	// IPA is the proof that shows that InnerProduct is correct
+	IPA *IPA
+}
+
+// rangeProver proves that a committed value < 2^BitLength.
+type rangeProver struct {
+	// value is the value committed in Commitment
+	value uint64
+	// blindingFactor is the randomness used to compute Commitment
+	blindingFactor *math.Zr
+	// Commitment is a hiding Pedersen commitment to value: Commitment = G^vH^r
+	Commitment *math.G1
+	// CommitmentGenerators are the generators (G, H) used to compute Commitment
+	CommitmentGenerators []*math.G1
+	// LeftGenerators are the generators that will be used to commit to
+	// the bits (b_0,..., b_{BitLength-1}) of value
+	LeftGenerators []*math.G1
+	// RightGenerators are the generators that will be used to commit to (b_i-1)
+	RightGenerators []*math.G1
+	// P is a random generator of G1
```

Comment:
I noticed that these public parameters are all injected through `setup.go`.
What's the reason for this? Why not instantiate them in this package instead globally?

Reply:
This package is used within the token library where all the public parameters are produced. As of now it is not used in a standalone capacity. If the need arises, this will be changed accordingly.